### PR TITLE
Move exp, log, derivative, integral from Misc to Series.jl, add tests and docs, etc.

### PR DESCRIPTION
**Pull request by wbhart**

### CODE REVIEW

**Comment by wbhart** [comment]
Also fixes a bug in addeq! for series.

**Comment by fieker** [comment]
On Thu, Mar 11, 2021 at 03:29:06AM -0800, wbhart wrote:
> Do not merge as I still have a question about correctness in one case, and am having trouble finding a literature reference.
What is missing? The vague reference if I think Richard Brent
https://maths-people.anu.edu.au/~brent/pd/rpb045.pdf

**Comment by wbhart** [comment]
@fieker Only one of exp(log(f)) == f and log(exp(f)) == f seems to hold over a field (assuming f has the correct valuation of course).

I will look in Brent.

**Comment by wbhart** [comment]
I assume that log(x) == 0 for x in Q iff x = 1, thus both identities only hold for formal power series over Q when the constant term is 1.

**Comment by wbhart** [comment]
So I am wondering if we need some additional restriction on the input of one of the functions.

**Comment by wbhart** [comment]
By the way, the following will likely produce faster results. Not that I have time to implement it myself right now:

https://link.springer.com/content/pdf/10.1007/s40065-017-0166-4.pdf

**Comment by wbhart** [comment]
Actually, probably the problem is we need to divide out by the constant term and multiply the result by the log/exp whatever it is of that constant. Thus I think the current code is in fact incorrect.

**Comment by fieker** [comment]
On Thu, Mar 11, 2021 at 03:52:57AM -0800, wbhart wrote:
> Actually, probably the problem is we need to divide out by the constant term and multiply the result by the log/exp whatever it is of that constant. Thus I think the current code is in fact incorrect.
> 
"my" code was only ever done for 1+... in log and 0+... in exp, your fix
is correct, however, the number of rationals with rational log is
limited...

**Comment by fieker** [comment]
On Thu, Mar 11, 2021 at 03:52:57AM -0800, wbhart wrote:
> Actually, probably the problem is we need to divide out by the constant term and multiply the result by the log/exp whatever it is of that constant. Thus I think the current code is in fact incorrect.
> 
Given that exp/log are transcendental, I don't think that outside
 - real, complex
 - padic
 - power series
this matters. For exact rings as base ring: 1 and 0 are neccessary

**Comment by wbhart** [comment]
Sure but this is a generic implementation. It's not just for rationals.

**Comment by wbhart** [comment]
This should be good to go, but there is an outside chance I broke Nemo or Hecke. Unfortunately I can't test them at present due to https://github.com/Nemocas/Nemo.jl/issues/1015

**Comment by fieker** [comment]
On Thu, Mar 11, 2021 at 04:47:37AM -0800, wbhart wrote:
> Sure but this is a generic implementation. It's not just for rationals.
Only imprecise rings have exp and log for non-trivial input

**Comment by wbhart** [comment]
Yes, but I added test code for an imprecise ring and it failed. It's all fixed now.

File: docs/src/series.md [hunk 1]
```diff
@@ -482,8 +482,38 @@ julia> d = inv(b)
...
+(Univariate power series ring in x over Rationals, x + O(x^11))
+
+julia> f = 2 + x + 3x^3
+2 + x + 3*x^3 + O(x^10)
+
+julia> derivative(f)
+1 + 9*x^2 + O(x^9)
+
+julia> integral(f)
+2*x + 1//2*x^2 + 3//4*x^4 + O(x^11)
```
Comment (fingolfin) [inline docs/src/series.md:508]: How about moving these kinds of examples into the respective docstrings, then they are also available in the REPL help. Julia provides many such examples in the docstrings, and I think that's very beneficial for many users.

Comment (wbhart) [reply docs/src/series.md]: Done.

File: src/generic/RelSeries.jl [hunk 2]
```diff
@@ -1010,12 +1010,103 @@ function Base.sqrt(a::AbstractAlgebra.RelSeriesElem)
...
+Return the derivative of the power series $f$.
+
+julia> R, x = PowerSeriesRing(QQ, 10, "x")
+(Univariate power series ring in x over Rationals, x + O(x^11))
+
+julia> f = 2 + x + 3x^3
+2 + x + 3*x^3 + O(x^10)
+
+julia> derivative(f)
+1 + 9*x^2 + O(x^9)
```
Comment (thofma) [inline src/generic/RelSeries.jl:1031]:
````suggestion
```jldoctes
julia> R, x = PowerSeriesRing(QQ, 10, "x")
(Univariate power series ring in x over Rationals, x + O(x^11))

julia> f = 2 + x + 3x^3
2 + x + 3*x^3 + O(x^10)

julia> derivative(f)
1 + 9*x^2 + O(x^9)
```
````

Comment (thofma) [reply src/generic/RelSeries.jl:1031]: Somehow I cannot apply this suggestion myself, strange.

**Comment by thofma** [comment]
Will merge once the lights turn green. Then I will tag the new version.

**Comment by thofma** [comment]
Of course the doctest is complaining :(

Can you enable the rights for maintainers to push to your branch? Not sure why it is disabled, should be enabled by default. Then I can push this over the finish line.

**Comment by wbhart** [comment]
I am not sure where I find that,  but I'll try to enable that. I certainly haven't disabled it myself as far as I recall.

**Comment by wbhart** [comment]
I don't find such an option. The only relevant thing seems to be interaction limits, and these are all currently disable (i.e. no limits).

Can you tell me what I am looking for?

**Comment by wbhart** [comment]
It seems that you are not currently a maintainer of Nemocas/AbstractAlgebra. I tried to add you but it says there is no GitHub account with the name thofma

Not sure how to fix this.

**Comment by wbhart** [comment]
I think GitHub is having issues. I just went to try the exact same again to take a screenshot and show you, and now it recognises thofma. I am 100% sure I spelled it correctly the first time!

Anyhow, you are now an admin and can do whatever you need on the main project. However, I'm unable to find such options on my personal repo. I think those options are only for community projects maybe?

**Comment by thofma** [comment]
I think it is an option for PR, somewhere on the right hand side.

**Comment by wbhart** [comment]
It's enabled, but you were not a maintainer. You should be now (Admin even).

**Comment by wbhart** [comment]
thofma

**Comment by thofma** [comment]
https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork

**Comment by wbhart** [comment]
thofma2

This is what was the problem. No idea why.

**Comment by wbhart** [comment]
If that doesn't work, I can switch your role from Admin to Maintainer. But the one seems to supersede the other from what I can tell.

**Comment by thofma** [comment]
Did you try https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork? It is an option per PR. Should be somewhere on the right hand side.

**Comment by wbhart** [comment]
Yes, I believe there is something wrong with GitHub. Do you see the screenshot above where I show that this is (and always was enabled)? Are you saying you still cannot push to my branch, even as an Admin?

**Comment by thofma** [comment]
Ah, sorry, I looked at the wrong picture. Everything looks right, but I was not able to push. Mysterious git/GitHub things going on today.

**Pull request merged**